### PR TITLE
*: track the memory usage of IndexJoin more accurate

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/28650

Problem Summary:
The memory usage of `indexJoinBuilderHelper.buildTemplateRange` and `indexRangesToKVWithoutSplit` is not tracked. Thus we can not cancel the SQL which uses IndexJoin with huge `nextColRanges` even if it costs a lot of memory.

### What is changed and how it works?
After this commit, the SQL **can be cancelled** when reaching the memory quota.

### Check List

Tests 

- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

[test.txt](https://github.com/pingcap/tidb/files/7553357/test.txt)

Side effects

- [x] Performance regression: Consumes more CPU
But this regression can be ignored since the newly added memory tracking is a low-frequency action.

Documentation
N/A

### Release note

```release-note
Track the memory usage of IndexJoin more accurate.
```

### CODE REVIEW

File: distsql/request_builder.go [hunk 1]
```diff
@@ -642,16 +655,31 @@ func VerifyTxnScope(txnScope string, physicalTableID int64, is infoschema.InfoSc
 	return true
 }
 
-func indexRangesToKVWithoutSplit(sc *stmtctx.StatementContext, tids []int64, idxID int64, ranges []*ranger.Range) ([]kv.KeyRange, error) {
+func indexRangesToKVWithoutSplit(sc *stmtctx.StatementContext, tids []int64, idxID int64, ranges []*ranger.Range, interruptSignal *atomic.Value) ([]kv.KeyRange, error) {
 	krs := make([]kv.KeyRange, 0, len(ranges))
-	for _, ran := range ranges {
+	const step = 8
+	var memUsage int64 = 0
```
Comment (guo-shaoge) [inline distsql/request_builder.go:661]:
```suggestion
	var memUsage int64
```

File: executor/builder.go [hunk 2]
```diff
@@ -4091,6 +4094,9 @@ func buildKvRangesForIndexJoin(ctx sessionctx.Context, tableID, indexID int64, l
 		}
 	}
 
+	if len(tmpDatumRanges) != 0 {
+		memTracker.Consume(2 * int64(len(tmpDatumRanges)) * types.EstimatedMemUsage(tmpDatumRanges[0].LowVal, len(tmpDatumRanges[0].LowVal)))
```
Comment (wshwsh12) [inline executor/builder.go:4098]:
```suggestion
		memTracker.Consume(2 * types.EstimatedMemUsage(tmpDatumRanges[0].LowVal, len(tmpDatumRanges)))
```

**Review by guo-shaoge — APPROVED** [review]

File: executor/index_lookup_join.go [hunk 3]
```diff
@@ -519,6 +550,7 @@ func (iw *innerWorker) constructLookupContent(task *lookUpJoinTask) ([]*indexJoi
 		numRows := chk.NumRows()
 		for rowIdx := 0; rowIdx < numRows; rowIdx++ {
 			dLookUpKey, dHashKey, err := iw.constructDatumLookupKey(task, chkIdx, rowIdx)
+			iw.lookup.memTracker.Consume(types.EstimatedMemUsage(dLookUpKey, len(dLookUpKey)))
```
Comment (wshwsh12) [inline executor/index_lookup_join.go:553]:
```suggestion
			iw.lookup.memTracker.Consume(types.EstimatedMemUsage(dLookUpKey, 1))
```

File: executor/index_lookup_join.go [hunk 4]
```diff
@@ -222,6 +227,16 @@ func (e *IndexLookUpJoin) newInnerWorker(taskCh chan *lookUpJoinTask) *innerWork
...
+		lookup:        e,
+		memTracker:    memory.NewTracker(memory.LabelForIndexJoinInnerWorker, -1),
+	}
+	iw.memTracker.AttachTo(e.memTracker)
+	for _, ran := range copiedRanges {
+		// We should not consume this memory usage in `iw.memTracker`. The
+		// memory usage of inner worker will be reset the end of iw.handleTask.
+		// While the life cycle of this memory consumption exists throughout the
+		// whole active period of inner worker.
+		e.ctx.GetSessionVars().StmtCtx.MemTracker.Consume(2 * types.EstimatedMemUsage(ran.LowVal, len(ran.LowVal)))
```
Comment (wshwsh12) [inline executor/index_lookup_join.go:239]: ditto

**Review by wshwsh12 — APPROVED** [review]
LGTM

**Pull request merged**